src/init.cpp

```diff
@@ -1886,6 +1883,48 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
 
 bool StartIndexBackgroundSync(NodeContext& node)
 {
+    // Find the oldest block among all indexes.
+    // This block is used to verify that we have the required blocks' data stored on disk,
+    // starting from that point up to the current tip.
+    // indexes_start_block='nullptr' means "start from height 0".
+    std::optional<const CBlockIndex*> indexes_start_block;
+    std::string older_index_name;
+
+    ChainstateManager& chainman = *Assert(node.chainman);
+    for (auto index : node.indexes) {
+        // If the index best block is nullptr, means that we will start from the genesis block
+        if (indexes_start_block && !indexes_start_block.value()) continue; // So, nothing to do.
+
+        const IndexSummary& summary = index->GetSummary();
+        if (summary.synced) continue;
+
+        // Get the last common block between the index best block and the active chain
+        LOCK(::cs_main);
+        const CChain& active_chain = chainman.ActiveChain();
+        const CBlockIndex* pindex = chainman.m_blockman.LookupBlockIndex(summary.best_block_hash);
+        if (!active_chain.Contains(pindex)) {
+            pindex = active_chain.FindFork(pindex);
+        }
+
+        // Update start block
+        bool need_start_from_scratch = !indexes_start_block || !pindex;
```

Comment:
In commit "index: verify blocks data existence only once" (30b2511f39d32e29f9f05859aa8a97b84c22376b)

The variable name `need_start_from_scratch` seems misleading because it will be true even when not syncing from genesis. Would suggest simplifying and clarifying with:

```diff
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -1892,9 +1892,6 @@ bool StartIndexBackgroundSync(NodeContext& node)
 
     ChainstateManager& chainman = *Assert(node.chainman);
     for (auto index : node.indexes) {
-        // If the index best block is nullptr, means that we will start from the genesis block
-        if (indexes_start_block && !indexes_start_block.value()) continue; // So, nothing to do.
-
         const IndexSummary& summary = index->GetSummary();
         if (summary.synced) continue;
 
@@ -1906,11 +1903,10 @@ bool StartIndexBackgroundSync(NodeContext& node)
             pindex = active_chain.FindFork(pindex);
         }
 
-        // Update start block
-        bool need_start_from_scratch = !indexes_start_block || !pindex;
-        if (need_start_from_scratch || pindex->nHeight < indexes_start_block.value()->nHeight) {
+        if (!indexes_start_block || !pindex || pindex->nHeight < indexes_start_block.value()->nHeight) {
             indexes_start_block = pindex;
             older_index_name = summary.name;
+            if (!pindex) break; // Starting from genesis so no need to look for earlier block.
         }
     };
 
```

Reply:
Sure, applied. Thanks!